server/redeemer.go

```diff
@@ -0,0 +1,398 @@
+package server
+
+// generate the go bindings for github.com/livepeer/go-livepeer/net/redeemer.proto first !
+//go:generate mockgen -source github.com/livepeer/go-livepeer/net/redeemer.pb.go -destination github.com/livepeer/go-livepeer/net/redeemer_mock.pb.go -package net
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	gonet "net"
+	"net/url"
+	"sync"
+	"time"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/livepeer/go-livepeer/eth"
+	"github.com/livepeer/go-livepeer/net"
+	"github.com/livepeer/go-livepeer/pm"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+)
+
+const rpcTimeout = 8 * time.Second
+
+var cleanupLoopTime = 1 * time.Hour
+
+// Redeemer is the interface for a ticket redemption gRPC service
+type Redeemer interface {
+	net.TicketRedeemerServer
+	Start(host *url.URL) error
+	Stop()
+}
+
+type redeemer struct {
+	recipient   ethcommon.Address
+	subs        sync.Map
+	eth         eth.LivepeerEthClient
+	sm          pm.SenderMonitor
+	quit        chan struct{}
+	liveSenders sync.Map // ethCommon.Address => time.Time lastAccess
+}
+
+// NewRedeemer creates a new ticket redemption service instance
+func NewRedeemer(recipient ethcommon.Address, eth eth.LivepeerEthClient, sm pm.SenderMonitor) (Redeemer, error) {
+
+	if recipient == (ethcommon.Address{}) {
+		return nil, fmt.Errorf("must provide a recipient")
+	}
+
+	if eth == nil {
+		return nil, fmt.Errorf("must provide a LivepeerEthClient")
+	}
+
+	if sm == nil {
+		return nil, fmt.Errorf("must provide a SenderMonitor")
+	}
+
+	return &redeemer{
+		recipient: recipient,
+		eth:       eth,
+		sm:        sm,
+		quit:      make(chan struct{}),
+	}, nil
+}
+
+func (r *redeemer) Start(host *url.URL) error {
+	listener, err := gonet.Listen("tcp", host.String())
+	if err != nil {
+		return err
+	}
+	defer listener.Close()
+	if err != nil {
+		return err
+	}
+	// slice of gRPC options
+	// Here we can configure things like TLS
+	opts := []grpc.ServerOption{}
+	// var s *grpc.Server
+	s := grpc.NewServer(opts...)
+	defer s.Stop()
```

Comment:
By the time this defer statement executes, the server would've stopped already when `s.Serve()` returns. May make more sense to store the server in the struct so that `Stop()` can be called in `redeemer.Stop()` which would allow the server to be stopped when the program exits.

Reply:
fixed in 44044b6da3ffbf0afde1a98be1354eea884017e3